README
# MyRNG: A convenient random number generator

The MyRNG C++ classes provide two very efficient random number generators 
and several methods for generating random variates. You can choose between 
the Mersenne Twister MT19937a algorithm in the implementation of [Makoto
Matsumoto and Takuji Nishimura][1] and the well-equidistributed long-period 
linear generator [WELL1024a][2] implemented by Marton Morvai. Both generators are
essentially shuffled linear congruential random number generators; they should
not be used for cryptography applications but should otherwise cover most 
scientific needs. While the WELL1024a is slightly faster in the current 
implementation, the MT19937a has an astronomical recurrence time of 2^19937. 
Furthermore, the methods are provided to generate several random variates 
including uniform, Gaussian, Beta, and Gamma distributions. Generation of most
random variates is based on [Law and Kelton, 2000][3]. Generation of the Gamma 
distribution is based on [Marsagli and Tsang, 2000][4].

## References

1: http://www.math.sci.hiroshima-u.ac.jp/~m-mat/MT/emt.html "M. Matsumoto and T. Nishimura. Personal web pages."
2: http://dx.doi.org/10.1145/1132973.1132974 "F. Panneton et al.: Improved long-period generators based on linear recurrences
    modulo 2, ACM Transactions on Mathematical Software, 32(1), 1-16,2006."
3: http://dl.acm.org/citation.cfm?id=550113 "A.M. Law and W.D. Kelton: Simulation, modeling and analysis, Third Edition, 
    McGraw Hill, 2006."
4: http://dx.doi.org/10.1145/358407.358414 "G. Marsaglia and W.W. Tsang: A simple method for generating gamma variables, 
    ACM TOMS 26, 2000."

ChangeLog
2012-08-28 12:53:59 +0200 Gerd Zschaler 

	* Update README.md (HEAD, v1.5, origin/master, origin/HEAD, master)

2012-08-28 12:39:35 +0200 Gerd Zschaler 

	* Remove obsolete test makefile

2012-08-28 12:00:22 +0200 Gerd Zschaler 

	* Allow for header-only use

2012-08-28 11:44:26 +0200 Gerd Zschaler 

	* Fix wrong include path in pkgconfig data

2012-07-16 17:21:35 +0200 Gerd Zschaler 

	* Add author information and change log

2012-07-16 17:16:30 +0200 Gerd Zschaler 

	* Compile as shared libraries using GNU Autotools

2012-07-03 15:30:10 +0300 Gerd Zschaler 

	* Readme formatting, mention doxygen

2012-06-29 13:32:17 +0200 Gerd Zschaler 

	* Fix utility functions to work with long integer types (v1.4)

2012-06-29 13:18:22 +0200 Gerd Zschaler 

	* Add test for unsigned int IntFromTo()

2012-06-29 13:17:26 +0200 Gerd Zschaler 

	* Member templates for random integer functions

2012-06-29 12:50:57 +0200 Gerd Zschaler 

	* Use namespace std for time()

2012-05-23 12:45:15 +0200 Gerd Zschaler 

	* Inline template functions

2012-05-23 12:42:59 +0200 Gerd Zschaler 

	* Utility functions to retrieve random elements from containers

2012-03-16 14:18:24 +0100 Gerd Zschaler 

	* Fix typo, formatting

2011-11-08 16:43:11 +0100 Gerd Zschaler 

	* Add .PHONY rule

2011-11-08 10:40:58 +0100 Gerd Zschaler 

	* Update README

2011-11-07 20:53:46 +0100 Gerd Zschaler 

	* Readme

2011-11-07 20:51:23 +0100 Gerd Zschaler 

	* Update README.md

2011-11-07 20:48:45 +0100 Gerd Zschaler 

	* Readme

2011-11-07 17:45:05 +0100 Gerd Zschaler 

	* Fix Markdown formatting (v1.3)

2011-11-07 16:24:17 +0100 Gerd Zschaler 

	* Rename README

2011-11-07 16:23:20 +0100 Gerd Zschaler 

	* Markdown format for README

2011-11-07 16:10:57 +0100 Gerd Zschaler 

	* First commit (version 1.3 from svn)



myrng/WELLEngine.h
#ifndef WELLENGINE_H_
#define WELLENGINE_H_

#include <cassert>
#include <ctime>

namespace myrng
{

/**
 * The WELL1024a random number generator.
 *
 * The acronym WELL stands for "Well Equidistributed Long-period Linear". It
 * has a period of 2^1024, which should be enough for most simulation
 * tasks. It has better equidistribution properties and it recovers faster
 * from a "bad state", e.g. bad initialization, than the MT19937.
 * For more information see F. Panneton, P. L'Ecuyer, and M. Matsumoto,
 * "Improved Long-Period Generators Based on Linear Recurrences Modulo 2",
 * ACM Transactions on Mathematical Software, 32, 1 (2006), 1-16.
 *
 * @author Gerd Zschaler <gzschaler@googlemail.com>
 * @author Marton Morvai
 * @author Thilo Gross <gross@physics.org>
 *
 * @copyright This implementation of the WELL alogrithm is licensed under the
 * Creative Commons Attribution-NonCommercial 3.0 Unported License. To view
 * a copy of this license, visit http://creativecommons.org/licenses/by-nc/3.0/
 * or send a letter to Creative Commons, 444 Castro Street, Suite 900, Mountain
 * View, California, 94041, USA.
 */
class WELLEngine
{
public:
	WELLEngine()
	{
		// Long has to be at least 32 bits!
		assert(sizeof(long) >= 4);
		state_i = 0;
		seed((unsigned long) std::time(0));
	}
	/**
	 * Set random number generator seed
	 */
	void seed(unsigned long int s)
	{
		/*
		 * After the engine has been seeded, this function generates 50.000
		 * random numbers to let it recover at least a little bit from a
		 * bad initial state.
		 */
		seed_ = s;
		STATE[0] = s & 0xffffffffUL;
		for (int i = 1; i < 32; ++i)
		{
			STATE[i] =
					(1812433253UL * (STATE[i - 1] ^ (STATE[i - 1] >> 30)) + i);
			STATE[i] &= 0xffffffffUL;
			// for >32 bit machines
		}
		for (int j = 0; j < 10000; ++j)
			(*this)();
	}
	/**
	 * Get random number generator seed
	 */
	unsigned long int getSeed() const
	{
		return seed_;
	}
	/**
	 * Get the name of the random number generator engine
	 */
	const char* getName()
	{
		return "WELL1024a";
	}

protected:
	~WELLEngine()
	{
	}
	/// Produce a random number.
	double operator()()
	{
		/*
		 * This generator is an Implementation of WELL1024a
		 * (F. Panneton, P. L'Ecuyer, and M. Matsumoto,
		 * "Improved Long-Period Generators Based on Linear Recurrences Modulo 2",
		 * ACM Transactions on Mathematical Software, 32, 1 (2006), 1-16.)
		 */
		z0 = STATE[(state_i + 31) & 0x0000001fU];
		z1 = (STATE[state_i]) ^ STATE[(state_i + 3) & 0x0000001fU];
		z2 = (STATE[(state_i + 24) & 0x0000001fU]
				^ (STATE[(state_i + 24) & 0x0000001fU] << (19)))
				^ (STATE[(state_i + 10) & 0x0000001fU]
						^ (STATE[(state_i + 10) & 0x0000001fU] << (14)));

		STATE[state_i] = z1 ^ z2;
		STATE[(state_i + 31) & 0x0000001fU] = (z0 ^ (z0 << (11)))
				^ (z1 ^ (z1 << (7))) ^ (z2 ^ (z2 << (13)));

		state_i = (state_i + 31) & 0x0000001fUL;
		return ((double) STATE[state_i] * 2.32830643653869628906e-10);
	}

private:
	unsigned long int seed_; ///< the initial seed

	unsigned int z0, z1, z2, state_i;
	unsigned int STATE[32];
};

} /* namespace myrng */
#endif /* WELLENGINE_H_ */
